### PR TITLE
[AIRFLOW-4456] Add sub-classable BaseBranchOperator

### PR DESCRIPTION
**Pull request by chronitis**
### Jira

- [x] My PR addresses https://issues.apache.org/jira/browse/AIRFLOW-4456

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This adds an abstract BranchOperator, which implementors can use to
create concrete operators for branching workflows fully encapsulated in
a class, rather than needing to pass an external callable (or
self-reference) to BranchPythonOperator. Also adds tests for same.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

`tests/operators/test_branch_operator.py`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by ashb** [comment]
If you're going to add this the existing BranchingPython operator should use it!

**Comment by ashb** [comment]
An alternative would be for you to subclass the existiing PythonBranchOperator like this in your code base:

```python
class MyBranchOperator(PythonBranchOperator):
    def __int__(**kwargs):
        super().__init__(python_callable=self.choose_branch, **kwargs)

    def choose_branch(self):
        return 'task_a' # etc.
```

**Comment by chronitis** [comment]
I take the comment that this logic shouldn't be duplicated between this operator and `BranchPythonOperator` - I'll look at a better way of avoiding duplication.

The reason for not subclassing `BranchPythonOperator` was that creating a bound reference to the method in `__init__` seems to cause issues with templating:

```python
class MyBranch(BranchPythonOperator):
    template_fields = ('myvar', )
    def __init__(self, myvar, **kwargs):
       self.myvar = myvar
       super().__init__(python_callable=self.choose_branch)

    def choose_branch(self):
        # for reasons I'm not totally clear on, but I'm guessing the reference to `self` before `super()`
        # has returned creates a closure, any {{ templates }} in `self.myvar` here don't get evaluated
        return self.myvar
```

Less importantly, it seems a bit inelegant to need to pass a reference to the method like this.

**Comment by ashb** [comment]
myvar won't be templated until just before `execute()` is called - when were you looking at the value? In init or in choose_branch?

**Comment by ashb** [comment]
> I take the comment that this logic shouldn't be duplicated between this operator and `BranchPythonOperator` - I'll look at a better way of avoiding duplication.

Yes, that was what I meant.

**Comment by chronitis** [comment]
> myvar won't be templated until just before `execute()` is called - when were you looking at the value? In init or in choose_branch?

In `choose_branch`; if you inspect the operator via the web interface the _rendered template_, `myvar` comes out as expected (templates applied), but code in `choose_branch` (invoked from `BranchPythonOperator.execute()`) appears to see the the untemplated version.

**Comment by chronitis** [comment]
Here's a possible refactoring; moving the logic part into `SkipMixin` and building both `BranchOperator` and `BranchPythonOperator` off it. It's possible that `BranchOperator` is unnecessary now and users should just inherit `SkipMixin` directly in this logic.

(Tests pass except for the Kubernetes case which appears to have timed out).

**Comment by chronitis** [comment]
@ashb does this refactoring and documentation look reasonable?

**Comment by potiuk** [comment]
> @ashb does this refactoring and documentation look reasonable?

Ash is on holidays for the next 3 weeks but I might take over from here. I will take a look later today/tomorrow ok?

**Comment by chronitis** [comment]
@potiuk fine with me. I'm grateful for whoever amongst the committers is willing to take a look.

File: airflow/operators/branch_operator.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from airflow.models import BaseOperator, SkipMixin
+
+from typing import Union, Iterable, Dict
+
+
+class BranchOperator(BaseOperator, SkipMixin):
```
Comment (ashb) [inline airflow/operators/branch_operator.py]:
```suggestion
class BaseBranchOperator(BaseOperator, SkipMixin):
```

Just to make it a tiny bit clearer that this is intended to be sub-classed (i.e. not used in a DAG, similar to BaseOperator)

**Review by ashb — CHANGES_REQUESTED** [review]
Small naming change, but yes this looks good, thanks!

File: docs/concepts.rst [hunk 2]
```diff
@@ -581,6 +581,26 @@ For example:
 
   start_op >> branch_op >> [continue_op, stop_op]
 
+If you wish to implement your own operators with branching functionality,
+you can inherit from ``BranchOperator``, which behaves similarly to
```
Comment (ashb) [inline docs/concepts.rst]:
```suggestion
you can inherit from :class:`~airflow.operators.branch_operator.BaseBranchOperator`, which behaves similarly to
```

File: docs/concepts.rst [hunk 3]
```diff
@@ -581,6 +581,26 @@ For example:
...
+``BranchPythonOperator`` but expects you to provide an implementation of the
```
Comment (ashb) [inline docs/concepts.rst]:
```suggestion
:class:`~airflow.operators.python_operator.BranchPythonOperator` but expects you to provide an implementation of the
```

File: docs/concepts.rst [hunk 4]
```diff
@@ -581,6 +581,26 @@ For example:
...
+method ``choose_branch``. As with the callable for ``BranchPythonOperator``,
```
Comment (ashb) [inline docs/concepts.rst]:
```suggestion
method ``choose_branch``. As with the callable for BranchPythonOperator,
```

**Review by ashb — CHANGES_REQUESTED** [review]
Improve doc linking to classes too.

The suggestions as they are will make the line-wrapping messy, so probably tidy that up too :)

**Comment by ashb** [comment]
@chronitis If you are able to make the small changes suggested by Monday I can _probably_ get this in to the 1.10.4 release.

**Comment by chronitis** [comment]
@ashb I think that should cover all the changes you have requested. Thanks for getting back to this one.

**Review by ashb — APPROVED** [review]

**Comment by ashb** [comment]
(I edited your branch to fix the pylint warning)

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
Good work @chronitis

**Comment by ashb** [comment]
@chronitis Can you see if I've made some silly mistake on cherry-picking this back to the release branch -- it's failing on it's Python 2 tests. https://github.com/apache/airflow/commit/3fc191c1416ac85021539153fe9ed1449c15f393

Come the morning I'll pull this out of the branch to get a 1.10.4beta1 out but hopefully we can get it back in before we create the first release candidate.

**Comment by ashb** [comment]
Oh got it - str (py3) vs basesxtring (py2)

**Pull request merged**